mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala

```diff
@@ -0,0 +1,153 @@
+
+  setDefault(selectedIndices -> Array.empty[Int])
+
+  /** @group getParam */
+  def getSelectedIndices: Array[Int] = getOrDefault(selectedIndices)
+
+  /** @group setParam */
+  def setSelectedIndices(value: Array[Int]): this.type = set(selectedIndices, value)
+
+  /**
+   * An array of feature names to select features from a vector column.
+   * @group param
+   */
+  val selectedNames = new StringArrayParam(this, "selectedNames",
+    "An array of feature names to select features from a vector column")
+
+  setDefault(selectedNames -> Array.empty[String])
+
+  /** @group getParam */
+  def getSelectedNames: Array[String] = getOrDefault(selectedNames)
+
+  /** @group setParam */
+  def setSelectedNames(value: Array[String]): this.type = set(selectedNames, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Slice a dense vector with an array of indices.
+   */
+  private def selectColumns(indices: Array[Int], features: DenseVector): Vector = {
+    Vectors.dense(indices.map(features.apply))
+  }
+
+  /**
+   * Slice a sparse vector with a set of indices.
+   */
+  private def selectColumns(indices: Set[Int], features: SparseVector): Vector = {
+    Vectors.sparse(
+      indices.size, features.indices.zip(features.values).filter(x => indices.contains(x._1)))
+  }
+
+  private def getFeatureIndicesFromNames(inputAttr: AttributeGroup): Array[Int] = {
+    $(selectedNames).map(name => inputAttr.getAttr(name).index.get)
+  }
+
+  private def merge(xs: List[Int], ys: List[Int]): List[Int] = {
+    (xs, ys) match {
+      case (Nil, ys) => ys
+      case (xs, Nil) => xs
+      case (x :: xs1, y :: ys1) =>
+        if (x < y) x :: merge(xs1, ys)
+        else y :: merge(xs, ys1)
+    }
+  }
+
+  /**
+   * Union feature indices from user specified indices and indices that transformed from user
+   * specified feature names. After the union process, indices are sorted ASC and any potential
+   * duplicates are removed.
+   */
+  private def unionFeatureIndices(first: Array[Int], second: Array[Int]): Array[Int] = {
+    merge(first.sorted.toList, second.sorted.toList).distinct.toArray
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    transformSchema(dataset.schema)
+
+    val indices = $(selectedIndices)
+    val slicer = udf { vec: Vector =>
+      vec match {
+        case features: DenseVector => selectColumns(indices, features)
+        case features: SparseVector => selectColumns(indices.toSet, features)
+      }
+    }
+    dataset.withColumn($(outputCol), slicer(dataset($(inputCol))))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    SchemaUtils.checkColumnType(schema, $(inputCol), new VectorUDT)
+    val inputAttr = AttributeGroup.fromStructField(schema($(inputCol)))
+    $(selectedNames).foreach { name =>
```

Comment:
These names (or any metadata) may not be available until previous parts of the Pipeline are fitted.  Can you please move this check to inside transform()?